File: 3903-smallest-stable-index-i/3903-smallest-stable-index-i.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var firstStableIndex = function (nums, k) {
    const n = nums.length
    let max = nums[0];
    let min = new Array(n).fill(0);

    min[n - 1] = nums[n - 1];
    for (let i = n - 2; i >= 0; i--) {
        min[i] = Math.min(min[i + 1], nums[i]);
    }

    for (let i = 0; i < nums.length; i++) {
        max = Math.max(max, nums[i]);

        if (max - min[i] <= k) {
            return i;
        }
    }

    return -1;
};

File: 3903-smallest-stable-index-i/3903-smallest-stable-index-i.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        mx = nums[0]
        mn = mx
        ans = -1

        for i in range(len(nums)):
            mx = max(mx, nums[i])
            mn = nums[i]
            for j in range(i, len(nums)):
                mn = min(mn, nums[j])

            if mx - mn <= k:
                ans = i
                break

        return ans
